### ISSUE TITLE
[setup] Check if boost python is installed

### ISSUE THREAD

**Issue by achilleas-k** [G-Node/nixpy#211]
Check for boost python after NIX is found and don't attempt to compile bindings if it's not found.

**Comment by s0nskar** [comment]
Hey, I want to take that. If anyone can drop any clues on how to do that it would be great.

**Comment by achilleas-k** [comment]
The main issue here is that the `check_nix` function (see [scripts/checknix.py](https://github.com/G-Node/nixpy/blob/ff8a923921441e2377ebe72f89101a3e1ee67c24/scripts/checknix.py#L11)) isn't that great (it's actually quite stupid).

All it does, is is creates a a cpp file which includes `<nix.hpp>` and tries to compile it. If it succeeds, then `setup.py` just goes ahead and starts compiling the bindings.

If boost-python isn't installed on the system, this will cause an error.

The best fix for this issue is a general way to check if all the (C++) dependencies are present, so both nix and boost-python.  Issue #209 is also relevant here, in that we generally want dependencies and dependency versions to be checked before attempting to build the bindings.

For starters though, we could also accept a general, cross-platform solution that simply tells us if boost-python is available on the system.

**Comment by s0nskar** [comment]
*how to check libboost_python is installed or which version is installed*

I searched a bit and found only this
```bash
s0nskar@a:~/code/nixpy$ ldconfig -p | grep libboost_python

libboost_python-py35.so.1.58.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libboost_python-py35.so.1.58.0
libboost_python-py35.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libboost_python-py35.so
libboost_python-py27.so.1.58.0 (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libboost_python-py27.so.1.58.0
libboost_python-py27.so (libc6,x86-64) => /usr/lib/x86_64-linux-gnu/libboost_python-py27.so
```
Now i have to parse the output, Is it the right way ?

**Comment by msuzen** [comment]
It could be searched from python programmatically using `apt` module in case OS is a debian base.
If  `pack_installed` in the below code has a size larger than 0 that means there is some `boost-python` package is installed. 

```python
import apt # use apt-get to install somehow pip has an issue
import apt_pkg
apt_pkg.init_config()
apt_pkg.init_system()
cache  = apt_pkg.Cache()                    # all cache packages
ipacks = [pack for pack in cache.packages] 
# Matches for starting with liboost-python
import re
pattern = re.compile(r'^libboost-python')
cached = [pack for pack in ipacks if pattern.match(pack.name)]
# Check if any is installed
pack_installed = [pack for pack in cached if pack.current_state == apt_pkg.CURSTATE_INSTALLED]
# Details if there is any installed
if len(pack_installed) > 0:
        # full name, version string
        print "installed names and version strings"
        [(pack.name, pack.current_ver.ver_str) for pack in pack_installed] 
```

**Comment by gicmo** [comment]
For finding boosts `BoostPyLib` in [findboost.py](https://github.com/G-Node/nixpy/blob/master/scripts/findboost.py) should do the trick on all platforms (and distributions) - if not we need to fix it.
For finding nix, we could use the package config file that we ship with nix `nix.pc` at least on unix.

**Comment by s0nskar** [comment]
thanks @gicmo ans @msuzen.
I think it would be better if we define a header like `#define NIX_LIB_VERSION "1.3.2"` like boost does in it's [boost/version.hpp](http://www.boost.org/doc/libs/1_55_0/boost/version.hpp). It would be better and Cross-platform solution for finding nix version. Any comments?

**Comment by s0nskar** [comment]
So here is a gist of what I'm gonna do.

First check for nix just as now by compiling the cpp code with `#include <nix.hpp>` and we can get the nix version from here if we define a macro as I mentioned above or otherwise we can get version by parsing `nix.pc` in Unix systems.

```cpp
    #include <nix.hpp>

    int main() {
        cout << NIX_LIB_VERSION; 
        return 0;
    }
```

Then we can check that boost-python is installed for currently running Python version. If boost-python is found, then we can check if `nixpy version == nix version` as mentioned in #209 and build the bindings. If any of cases failed we can stop the installation there.

Issue it gonna fix: #209, #211, #243.

**Comment by achilleas-k** [comment]
I opened this issue because I came across a situation where NIX was detected but it failed to build the bindings because it couldn't find boost-python. I can't reproduce this situation.

I can't find a way to get nix on a system without having boost. I'm starting to suspect that I'm misremembering what happened and this issue may be nonsensical.
